File: Old_xml_and_scripts/Jim/high_skills/new-18.5.2013/TurnToPosition.rb
require 'high_skills/ruby_high_skill.rb'
include_class Java::sk.fiit.robocup.library.geometry.Angles
include_class Java::java.lang.Math

class TurnToPosition < RubyHighSkill

  def TurnToPosition.angle_range from, to
    	(from/180.0*Math::PI)..(to/180.0*Math::PI)
  end
  	
  @@turn_right_range = angle_range(180.0, 360.0) #Math::PI...(Math::PI*2.0)
  @@turn_left_range = angle_range(0.0, 180.0) #0...Math::PI
  @@in_place_range1 = angle_range(0.0, 15.0) #0...(Math::PI*0.1)
  @@in_place_range2 = angle_range(345.0, 360.0) #(Math::PI*1.9)...(Math::PI*2.0)

  def initialize target, validity_proc
    super()
    @target = target

    @validity_proc = validity_proc || Proc.new{return true}
    @ending = false
    
    @agentInfo = Java::sk.fiit.jim.agent.AgentInfo.getInstance
  end

def pickLowSkill
 	  is_still_valid = @validity_proc.call
    @ending = true if not is_still_valid   
    return nil if not is_still_valid or @ending

    #@target = @agentInfo.kickTarget()
    me_flattened = my.position.set_z(0.0)
    target_vector = @target - me_flattened
     target_angle = target_vector.phi
     diff_against_current = Angles.normalize(target_angle - my.rotation_z)
     @target_phi = diff_against_current #@target.getPhi()
    
    
    case
      when me.on_ground? : nil
      when @@in_place_range1.include?(Angles.normalize(@target_phi)) : nil
      when @@in_place_range2.include?(Angles.normalize(@target_phi)) : nil
      when @@turn_left_range.include?(Angles.normalize(@target_phi)) : get_skill("step_circ_right_small")
      when @@turn_right_range.include?(Angles.normalize(@target_phi)) : get_skill("step_circ_left_small")
      else return get_skill("step_circ_left_small")
    end
  end

 # def Turn.is_towards? point
 #   me_flattened = my.position.set_z(0.0)
 #   target_vector = point - me_flattened
 #   target_angle = target_vector.phi
 #   diff_against_current = Angles.normalize(target_angle - my.rotation_z)
 #   @@in_place_range1.include?(diff_against_current) or @@in_place_range2.include?(diff_against_current)
 # end

  def checkProgress
  #  raise "Fallen" if me.on_ground?
  end
end

#in order not to defer the startup cost
TurnToPosition.new 0.0, nil

File: Jim/src/sk/fiit/jim/agent/highskill/TurnToPosition.java
package sk.fiit.jim.agent.highskill;

import sk.fiit.jim.LambdaCallable;
import sk.fiit.jim.agent.models.AgentModel;
import sk.fiit.jim.agent.moves.LowSkill;
import sk.fiit.jim.agent.moves.LowSkills;
import sk.fiit.jim.agent.skills.HighSkill;
import sk.fiit.robocup.library.geometry.Angles;
import sk.fiit.robocup.library.geometry.Vector3D;
/*
 * 
 * Reimplemented from Ruby to Java by Michal Blanarik
 */
public class TurnToPosition extends HighSkill{

	private Angles rightRange = new Angles(180.0, 360.0);	//Math::PI...(Math::PI*2.0)
	private Angles leftRange = new Angles(0.0, 180.0);		//0...Math::PI
	private Angles inPlaceRange1 = new Angles(0.0, 15.0);	//0...(Math::PI*0.1)
	private Angles inPlaceRange2 = new Angles(345.0, 360.0);//(Math::PI*1.9)...(Math::PI*2.0)
	private boolean ending = false;
	
	private LambdaCallable validityProc;
	
	Vector3D target;
	
	public TurnToPosition(Vector3D target, LambdaCallable validityProc ) {
        this.target = target;
        this.validityProc = validityProc;
    }
	
	@Override
	public LowSkill pickLowSkill() {
		boolean isStillValid = this.validityProc.call();
		if(!isStillValid){
			this.ending = true;
		}
		if(!isStillValid || this.ending){
			return null;
		}
				
		
		AgentModel agentModel = AgentModel.getInstance();
		
		Vector3D me_flattened = agentModel.getPosition().setZ(0.0); 
		Vector3D target_vector = target.subtract(me_flattened);
		Double target_angle = target_vector.getPhi();
		Double diff_against_current = Angles.normalize(target_angle - agentModel.getRotationZ());
		
		Double target_phi = diff_against_current; 
		
		// On ground, high skill ends
        if (agentModel.isOnGround() || agentModel.isLyingOnBack() || agentModel.isLyingOnBelly()) {
            return null;
        } // Considered straight to target, high skill ends
        else if (inPlaceRange1.include(Angles.normalize(target_phi))) {
            return null;
        } // Considered straight to target, high skill ends
        else if (inPlaceRange2.include(Angles.normalize(target_phi))) {
            return null;
        } else if (leftRange.include(Angles.normalize(target_phi))) {
            return LowSkills.get("test_turn_left");
        } else if (rightRange.include(Angles.normalize(target_phi))) {
            return LowSkills.get("test_turn_right");
        } else {
            return LowSkills.get("test_turn_right");
        }
	}

	@Override
	public void checkProgress() {
	}

}
